virtual_rainforest/core/axes.py

```diff
@@ -0,0 +1,396 @@
+
+    def can_validate(  # noqa: D102
+        self, value: DataArray, grid: Grid, **kwargs: Any
+    ) -> bool:
+        return self.dim_names.issubset(value.dims) and not self.dim_names.issubset(
+            value.coords
+        )
+
+    def run_validation(  # noqa: D102
+        self, value: DataArray, grid: Grid, **kwargs: Any
+    ) -> DataArray:
+
+        # Cell ID is only a dimenson with a give length - assume the order correct and
+        # check the right number of cells found
+        n_found = value["cell_id"].size
+        if grid.n_cells != n_found:
+            raise ValueError(
+                f"Grid defines {grid.n_cells} cells, data provides {n_found}"
+            )
+
+        return value
+
+
+class Spat_XY_Coord_Square(AxisValidator):
+    """Spatial validator for XY coordinates onto a square grid.
+
+    This spatial axis validator applies to a  DataArray that has a x and y dimensions
+    with valued coordinates, which should map onto the grid cell ids, allowing for a
+    subset of ids.
+    """
+
+    core_axis = "spatial"
+    dim_names = {"x", "y"}
+
+    def can_validate(  # noqa: D102
+        self, value: DataArray, grid: Grid, **kwargs: Any
+    ) -> bool:
+        return self.dim_names.issubset(value.dims) and self.dim_names.issubset(
+            value.coords
+        )
+
+    def run_validation(  # noqa: D102
+        self, value: DataArray, grid: Grid, **kwargs: Any
+    ) -> DataArray:
+
+        # Get x and y coords to check the extents and cell coverage.
+        #
+        # TODO - Note that mapping all the cells here is a bit extreme with a square
+        # grid - could just map one row and column to confirm the indexing into the
+        # data, but this is a more general solution.
+
+        # Use .stack() to convert the axis into stacked pairs of values
+        xy_pairs = value.stack(cell=("y", "x")).coords
+        idx_pairs = value.drop_vars(("x", "y")).stack(cell=("y", "x")).coords
+
+        # Get the mapping of points onto the grid
+        idx_x, idx_y = grid.map_xy_to_cell_indexing(
+            x_coords=xy_pairs["x"].values,
+            y_coords=xy_pairs["y"].values,
+            x_idx=idx_pairs["x"].values,
+            y_idx=idx_pairs["y"].values,
+            strict=False,
+        )
+
+        # TODO - fine a way to enable strict = True - probably just kwargs.
+
+        # Now remap the grids from xy to cell_id - this uses the rather under described
+        # vectorized indexing feature in xarray:
+        #
+        #   https://docs.xarray.dev/en/stable/user-guide/indexing.html
+        #
+        # Specifically using DataArray.isel to select indexed dimensions by name, which
+        # avoids issues with different permutations of the axes - and map XY onto a
+        # common cell_id
+
+        return value.isel(
+            x=DataArray(idx_x, dims=["cell_id"]), y=DataArray(idx_y, dims=["cell_id"])
+        )
+
+
+class Spat_XY_Dim_Square(AxisValidator):
```

Comment:
Distinction between this validator and the one above should definitely be made clearer somewhere